File: platform/android/MapLibreAndroid/src/sharedRenderer/opengl/java/org/maplibre/android/maps/renderer/OpenGLRendererStrategy.java
package org.maplibre.android.maps.renderer;

import android.content.Context;
import android.view.Surface;
import android.view.TextureView;

import androidx.annotation.NonNull;

import org.maplibre.android.maps.renderer.surfaceview.GLSurfaceViewMapRenderer;
import org.maplibre.android.maps.renderer.surfaceview.MapLibreGLSurfaceView;
import org.maplibre.android.maps.renderer.surfaceview.SurfaceViewMapRenderer;
import org.maplibre.android.maps.renderer.textureview.GLTextureViewRenderThread;
import org.maplibre.android.maps.renderer.textureview.TextureViewMapRenderer;

/**
 * OpenGL concrete impl behind {@link MapRendererFactory}. Lives alongside the
 * other GL renderer helpers in src/sharedRenderer/opengl/.
 */
final class OpenGLRendererStrategy {

  private OpenGLRendererStrategy() {}

  static void attachTextureRenderThread(@NonNull TextureView textureView,
                                        @NonNull TextureViewMapRenderer renderer) {
    renderer.setRenderThread(new GLTextureViewRenderThread(textureView, renderer));
  }

  static SurfaceViewMapRenderer createSurfaceViewRenderer(@NonNull Context context,
                                                          String localFontFamily,
                                                          boolean renderSurfaceOnTop,
                                                          Runnable initCallback) {
    MapLibreGLSurfaceView surfaceView = new MapLibreGLSurfaceView(context);
    surfaceView.setZOrderMediaOverlay(renderSurfaceOnTop);
    return new GLSurfaceViewMapRenderer(context, surfaceView, localFontFamily) {
      @Override
      public void onSurfaceCreated(Surface surface) {
        initCallback.run();
        super.onSurfaceCreated(surface);
      }
    };
  }
}

File: platform/android/MapLibreAndroid/src/sharedRenderer/opengl/java/org/maplibre/android/maps/renderer/OpenGLRendererStrategy.kt
package org.maplibre.android.maps.renderer

import android.content.Context
import android.view.Surface
import android.view.TextureView
import org.maplibre.android.maps.renderer.surfaceview.GLSurfaceViewMapRenderer
import org.maplibre.android.maps.renderer.surfaceview.MapLibreGLSurfaceView
import org.maplibre.android.maps.renderer.surfaceview.SurfaceViewMapRenderer
import org.maplibre.android.maps.renderer.textureview.GLTextureViewRenderThread
import org.maplibre.android.maps.renderer.textureview.TextureViewMapRenderer

/**
 * OpenGL concrete impl behind [MapRendererFactory]. Lives alongside the
 * other GL renderer helpers in src/sharedRenderer/opengl/.
 */
internal object OpenGLRendererStrategy {
    fun attachTextureRenderThread(
        textureView: TextureView,
        renderer: TextureViewMapRenderer,
    ) {
        renderer.setRenderThread(GLTextureViewRenderThread(textureView, renderer))
    }

    fun createSurfaceViewRenderer(
        context: Context,
        localFontFamily: String?,
        renderSurfaceOnTop: Boolean,
        initCallback: Runnable,
    ): SurfaceViewMapRenderer {
        val surfaceView = MapLibreGLSurfaceView(context)
        surfaceView.setZOrderMediaOverlay(renderSurfaceOnTop)
        return object : GLSurfaceViewMapRenderer(context, surfaceView, localFontFamily) {
            override fun onSurfaceCreated(surface: Surface?) {
                initCallback.run()
                super.onSurfaceCreated(surface)
            }
        }
    }
}
